src/rdpc/findMatchedTNPairs.ts

```diff
@@ -0,0 +1,99 @@
+import _ from "lodash";
+import {
+  DonorData,
+  SamplePair,
+  StringMap,
+  TumourNormalDesignationValue,
+} from "rdpc/types";
+
+// iterates over donor's DonorData and finds all the matched
+// tumour/normal sample pairs. Matched samples must have the same
+// experiment strategy; tumour sample uses
+// matchedNormalSubmitterSampleId field as a foreign key
+// to look for the matched normal sample.
+export const findMatchedTNPairs = (
+  map: StringMap<DonorData>
+): StringMap<SamplePair[]> => {
+  const donorsWithSamplePairs: StringMap<SamplePair[]> = {};
+
+  Object.entries(map).forEach(([donorId, donorData]) => {
+    const allSamplePairs: SamplePair[] = [];
+    donorData.samplePairs.forEach((currentSample) => {
+      if (currentSample.tumourSample) {
+        const matchedNormalSample = donorData.samplePairs.filter(
+          (sample) =>
+            sample.normalSample &&
+            sample.normalSample.tumourNormalDesignation ===
+              TumourNormalDesignationValue.Normal &&
+            !sample.tumourSample &&
+            sample.normalSample.experimentStrategy ===
+              currentSample.tumourSample?.experimentStrategy &&
+            sample.normalSample.submitterSampleId ===
+              currentSample.tumourSample?.matchedNormalSubmitterSampleId
+        );
+
+        matchedNormalSample.forEach((matched) => {
+          const samplePair = {
+            normalSample: matched.normalSample,
+            tumourSample: currentSample.tumourSample,
+            firstPublishedAt: determinSamplePairFirstPublishedDate(
```

Comment:
You can probably just move the logic from this method right into this forEach loop, and instead of throwing an error just don't add the samplePair to the array if there is an issue with one of the samples.